src/realm/query.cpp

```diff
@@ -970,29 +971,58 @@ size_t Query::find_best_node(ParentNode* pn) const
 void Query::aggregate_internal(ParentNode* pn, QueryStateBase* st, size_t start, size_t end,
                                ArrayPayload* source_column) const
 {
+    auto current_node = pn->m_children[st->m_best_node];
     while (start < end) {
-        // Executes start...end range of a query and will stay inside the condition loop of the node it was called
-        // on. Can be called on any node; yields same result, but different performance. Returns prematurely if
-        // condition of called node has evaluated to true local_matches number of times.
-        // Return value is the next row for resuming aggregating (next row that caller must call aggregate_local on)
-        size_t best = find_best_node(pn);
-        start = pn->m_children[best]->aggregate_local(st, start, end, findlocals, source_column);
-
-        // Make remaining conditions compute their m_dD (statistics)
-        for (size_t c = 0; c < pn->m_children.size() && start < end; c++) {
-            if (c == best)
-                continue;
-
-            // Skip test if there is no way its cost can ever be better than best node's
-            double cost = pn->m_children[c]->cost();
-            if (pn->m_children[c]->m_dT < cost) {
+        if (st->m_local_match_count >= findlocals) {
+            // Time to evaluate other nodes
+            // Update m_dD
+            current_node->m_dD = st->m_number_checked / st->m_local_match_count;
+            double current_cost = current_node->cost();
+            bool re_evaluate = false;
+
+            // Make remaining conditions compute their m_dD (statistics)
+            for (size_t c = 0; c < pn->m_children.size() && start < end; c++) {
+                if (c == st->m_best_node)
+                    continue;
+
+                // Skip test if there is no way its cost can ever be better than best node's
+                if (pn->m_children[c]->m_dT > current_cost)
+                    continue;
 
                 // Limit to bestdist in order not to skip too large parts of index nodes
-                size_t maxD = pn->m_children[c]->m_dT == 0.0 ? end - start : bestdist;
-                size_t td = pn->m_children[c]->m_dT == 0.0 ? end : (start + maxD > end ? end : start + maxD);
-                start = pn->m_children[c]->aggregate_local(st, start, td, probe_matches, source_column);
+                size_t td = pn->m_children[c]->m_dT == 0.0 ? end : start + bestdist;
+                size_t prev_pos = start;
+                st->m_local_match_count = 0;
+                start =
+                    pn->m_children[c]->aggregate_local(st, start, std::min(td, end), probe_matches, source_column);
+
+                // Calculate new m_dD
+                if (st->m_local_match_count > 0) {
```

Comment:
How about just dividing by st->m_local_match_count+1 instead of the if-then-else?

Reply:
I did consider that, and now that you make the suggestion, I will make that simplification.